src/xrpld/app/tx/detail/InvariantCheck.cpp

```diff
@@ -2170,4 +2179,887 @@ ValidAMM::finalize(
+                    {
+                        JLOG(j.fatal()) <<  //
+                            "Invariant failed: set must not change assets "
+                            "available";
+                        result = false;
+                    }
+
+                    if (beforeShares && updatedShares &&
+                        beforeShares->sharesTotal != updatedShares->sharesTotal)
+                    {
+                        JLOG(j.fatal()) <<  //
+                            "Invariant failed: set must not change shares "
+                            "outstanding";
+                        result = false;
+                    }
+
+                    return result;
+                }
+                case ttVAULT_DEPOSIT: {
+                    bool result = true;
+
+                    auto const vaultDeltaAssets =
+                        deltaAssets(afterVault_->pseudoId);
+
+                    if (!vaultDeltaAssets)
+                    {
+                        JLOG(j.fatal()) <<  //
+                            "Invariant failed: deposit must change vault "
+                            "balance";
+                        return false;  // That's all we can do
+                    }
+
+                    if (*vaultDeltaAssets > tx[sfAmount])
+                    {
+                        JLOG(j.fatal()) <<  //
+                            "Invariant failed: deposit must not change "
+                            "vault balance by more than deposited amount";
+                        result = false;
+                    }
+
+                    if (*vaultDeltaAssets <= zero)
+                    {
+                        JLOG(j.fatal()) <<  //
+                            "Invariant failed: deposit must increase "
+                            "vault balance";
+                        result = false;
+                    }
+
+                    // Any payments (including deposits) made by the issuer
+                    // do not change their balance, but create funds instead.
+                    bool const issuerDeposit = [&]() -> bool {
+                        if (vaultAsset.native())
+                            return false;
+                        return tx[sfAccount] == vaultAsset.getIssuer();
+                    }();
+
+                    if (!issuerDeposit)
+                    {
+                        auto const accountDeltaAssets =
+                            [&]() -> std::optional<Number> {
+                            auto ret = deltaAssets(tx[sfAccount]);
+                            // Compensate for transaction fee deduced from
+                            // sfAccount
+                            if (ret && vaultAsset.native())
+                                *ret += fee.drops();
+                            if (ret && *ret == zero)
+                                return std::nullopt;
+                            return ret;
+                        }();
```

Comment:
Don't need to check `ret` twice. Maybe something like this?

       auto const accountDeltaAssets =
            [&]() -> std::optional<Number> {
            if (auto ret = deltaAssets(tx[sfAccount]))
            {
                // Compensate for transaction fee deduced from
                // sfAccount
                if (vaultAsset.native())
                    *ret += fee.drops();
                if (*ret != zero)
                    return ret;
            }
            return std::nullopt;
        }();